### PR TITLE
raftstore: support to switch in raft-engine and raft-db

### PR DESCRIPTION
**Pull request by hicqu**
Signed-off-by: qupeng <qupeng@pingcap.com>

### What problem does this PR solve?

Support to switch in raft-engine and raft-db.

### Check List 

Tests 

- Integration test

### Release note 
* No relrease note.

### CODE REVIEW

File: src/config.rs [hunk 1]
```diff
@@ -2383,29 +2384,18 @@ impl TiKvConfig {
...
+            && !RocksEngine::exists(&self.raft_store.raftdb_path)
+            && !RaftLogEngine::exists(&self.raft_engine.config.dir)
+        {
+            return Err("default rocksdb exist, but raftdb and raft engine not exist".into());
+        }
+        if !RocksEngine::exists(&kv_db_path)
+            && (RocksEngine::exists(&self.raft_store.raftdb_path)
+                || RaftLogEngine::exists(&self.raft_engine.config.dir))
+        {
+            return Err("default rocksdb not exist, but raftdb or raft engine exist".into());
```
Comment (Connor1996) [inline src/config.rs:2398]:
```suggestion
            return Err("default rocksdb doesn't exist, but raftdb or raft engine exist".into());
```

File: src/config.rs [hunk 2]
```diff
@@ -2383,29 +2384,18 @@ impl TiKvConfig {
...
-                && RaftLogEngine::exists(&self.raft_engine.config.dir)
-            {
-                return Err("default rocksdb not exist, but raft engine exist".into());
-            }
+
+        if RocksEngine::exists(&kv_db_path)
+            && !RocksEngine::exists(&self.raft_store.raftdb_path)
+            && !RaftLogEngine::exists(&self.raft_engine.config.dir)
+        {
+            return Err("default rocksdb exist, but raftdb and raft engine not exist".into());
```
Comment (Connor1996) [inline src/config.rs:2392]:
```suggestion
            return Err("default rocksdb exist, but both raftdb and raft engine don't exist".into());
```

File: components/server/src/raft_engine_switch.rs [hunk 3]
```diff
@@ -0,0 +1,355 @@
...
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use engine_rocks::raw::DBOptions;
+
+    fn do_test_switch(custom_raft_db_wal: bool) {
```
Comment (Connor1996) [inline components/server/src/raft_engine_switch.rs:264]: add a case that dump is aborted in the half way.

File: components/server/src/raft_engine_switch.rs [hunk 4]
```diff
@@ -0,0 +1,418 @@
...
+    }
+    engine.sync().unwrap();
+    info!(
+        "Finished dump, total regions: {}; Total bytes: {}; Consumed time: {:?}",
+        count_region,
+        count_size.load(Ordering::Relaxed),
+        consumed_time.elapsed(),
+    );
+
+    rename_to_tmp_dir(&raft_engine_path, &dirty_raft_engine_path);
```
Comment (Connor1996) [inline components/server/src/raft_engine_switch.rs:265]: what if raft-engine dir is same as raftdb dir

Comment (hicqu) [reply components/server/src/raft_engine_switch.rs:265]: Now it's checked in config.rs.

**Review by Connor1996 — APPROVED** [review]
LGTM

**Review by hunterlxt — APPROVED** [review]
/LGTM

**Pull request merged**